tools/topology/v2/attributes.conf

```diff
@@ -0,0 +1,55 @@
+# common attributes and attribute types. Attributes, defined here can then be
+# used by any blocks. E.g. the "format" attribute can be used by pipelines,
+# DAIs, etc.
+
+# we're using an "availability" parameter in attribute declaration instead of
+# adding ".optional" or ".mandatory" to the array name
+
+DefineType.format {
+	enum [
+		"s16le"
+		"s24le"
+		"s32le"
+		"float"
+	]
+}
+
+DefineType.time_domain {
+	enum [
+		"dma"
+		"timer"
+	]
+}
+
+DefineAttribute.tokens {
```

Comment:
This is a "tokens" *attribute* for cases where higher level configuration files (e.g. machine level) pass their specific token *arrays* to lower level configuration files (pipelines). This is currently used for example to define pipelines:
```
dnl W_PIPELINE(stream, period, priority, core, initiator, platform)
							  ^^^^^^^^
define(`W_PIPELINE',
...
`SectionWidget."'N_PIPELINE($1)`" {'
...
`	data ['
`		"'N_PIPELINE($1)`_data"'
`		"'$6`"'
		^^^^^^
`	]'
`}')
```
which then is called as
```
W_PIPELINE(N_DAI_OUT, SCHEDULE_PERIOD, SCHEDULE_PRIORITY, SCHEDULE_CORE, SCHEDULE_TIME_DOMAIN, pipe_dai_schedule_plat)
											       ^^^^^^^^^^^^^^^^^^^^^^
```
and `pipe_dai_schedule_plat` is an *array* of tokens, defined in bxt.m4 or similar. So far I don't have a good idea how to avoid such token array passing. The only thing we can make easier is make arrays of word- and string-tokens a built-in type, without this defining them directly as a structure of an array of string tokens and an array of word tokens becomes really painful.

Reply:
But we are not using SectionWidget or token array directly - we abstract all tokens with Attributes.

Reply:
@lgirdwood we do that where possible, yes. E.g. for volume, where it's known in advance which tokens must be supplied. But here it can be any tokens that the respective platform decides to apply. So I don't see how we can replace them with fixed attributes?